src/GafferOSL/ShadingEngine.cpp

```diff
@@ -191,6 +195,21 @@ class RenderState
 
 		bool userData( ustring name, TypeDesc type, void *value ) const
 		{
+			if( name == gIndex )
+			{
+				// if a 4 byte type has been requested then ensure we fit and cast to narrower type
+				// this way f32 reads of _index will succeed.
+				if( type.size() == sizeof( int ) && m_pointIndex <= ( (size_t) std::numeric_limits<int>::max() ) )
+				{
+					int v = ( int ) m_pointIndex;
+					return ShadingSystem::convert_value( value, type, &v, OIIO::TypeDesc( OIIO::TypeDesc::INT32 ) );
+				}
+				else
+				{
+					return ShadingSystem::convert_value( value, type, &m_pointIndex, OIIO::TypeDesc( OIIO::TypeDesc::UINT64 ) );
```

Comment:
OSL docs seem to hedge and say that ints are "at least 32 bit", but in the implementation it looks like they _are_ 32 bit, and looking at the source for convert_value, it doesn't appear to support anything else. The `getattribute()` return value could be used to detect failure from a shader, but I doubt anyone will check it for something they think of as built-in. I don't think we have an N-times functionality in the IECore::MessageHandler, but if we did, it wouldn't affect how you'd output the message - it'd be filtered after the fact.

I'm nitpicking really - just didn't like to have a code path that looks functional but isn't. Maybe we just add a comment explaining the situation and merge?